blackboxopt/optimizers/staged/bohb.py

```diff
@@ -125,7 +131,9 @@ def convert_from_statsmodels_kde_representation(
 
 
 def impute_conditional_data(
-    array: np.ndarray, vartypes: Union[list, np.ndarray]
+    array: np.ndarray,
+    vartypes: Union[list, np.ndarray],
+    rng: Optional[np.random.Generator] = None,
```

Comment:
Is there a particular reason to be conservative here, and not allow integer seeds? Because the use in the function suggest that this could be the default numpy seed type with everything it allows.

Reply:
> Is there a particular reason to be conservative here, and not allow integer seeds?

Yes, my motivation was, that re-seeding a function/logic that's called repeatedly could result in less random behavior. E.g. if `impute_conditional_data()` would be called with an integer seed instead of a Generator, I would have to ensure to input a _different_ seed at every call to ensure "real" random behavior across multiple calls of that function. I hope to avoid that be using a single RNG throughout the whole optimizer.

Does this make sense?

Reply:
I definitely see how in the current use case all we (should) do is pass around a single RNG, but the function signature itself doesn't need to know that about its use, I'd say. My proposal would be to make the type hint as generic as it can be and still use the function in exactly the way we do with the current proposal.

Reply:
What do you see as benefit of having a generic type hint here?

I'd say a more precise type hint discourage inappropriate usage (e.g. passing an integer instead) a bit more.

Reply:
> What do you see as benefit of having a generic type hint here?

That it accurately represents all valid ways in which this function can be used.

> I'd say a more precise type hint discourage inappropriate usage (e.g. passing an integer instead) a bit more.

Which would be inappropriate in the current context it is used. I guess the difference in our view points is how much one considers reusability. Because if one only considers the only current use, then having the restrictive type hint can be appropriate. If one considers reusability, it's unnecessarily restrictive because the inappropriateness of passing an integer solely comes from the context and not the function itself. There can be call contexts where it's ok to use an int instead of an rng.
So we can either call scope creep or premature optimization, and there are valid arguments for both, none of which warrant delaying this PR for longer in my view :relaxed:

Reply:
Thanks for your explanation. I'll leave it as is for now, then and we can adjust the type hint once we reuse the methods in a situation where it really would be appropriate to call them with another type then the specified one.